libraries/lib-graphics/PixelSampleMapper.h

```diff
@@ -0,0 +1,69 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  PixelSampleMapper.h
+
+  Dmitry Vedenko
+
+**********************************************************************/
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <variant>
+
+// Clang will fail to instantiate a variant if sampleCount is forward declared
+// It tries to instantiate std::invoke_result for some reason
+#include "SampleCount.h"
+
+class GRAPHICS_API PixelSampleMapper final
+{
+public:
+   PixelSampleMapper() = default;
+   PixelSampleMapper(const PixelSampleMapper&) = default;
+   PixelSampleMapper(PixelSampleMapper&&) = default;
+   PixelSampleMapper& operator=(const PixelSampleMapper&) = default;
+   PixelSampleMapper& operator=(PixelSampleMapper&&) = default;
+
+   PixelSampleMapper(double t0, double rate, double samplesPerPixel) noexcept;
+
+   void applyBias(double bias) noexcept;
+
+   double applyCorrection(
+      const PixelSampleMapper& oldMapper, size_t oldLen, size_t newLen);
+
+   sampleCount GetFirstSample(uint32_t column) const;
+   sampleCount GetLastSample(uint32_t column) const;
+   std::pair<sampleCount, sampleCount> GetSampleRange(uint32_t column) const;
+
+   using CustomMapper = std::function<sampleCount(uint32_t)>;
+   void setCustomMapper(CustomMapper mapper);
+
+   bool IsValid() const;
+   bool IsLinear() const noexcept;
+
+private:
+   struct LinearMapper final
+   {
+      // Fixes GCC7 build issues (constructor required before non-static data member)
+      LinearMapper() noexcept {}
+
+      LinearMapper(double initialValue, double samplesPerPixel) noexcept
+         : mInitialValue(initialValue)
+         , mSamplesPerPixel(samplesPerPixel)
+      {}
+
+      LinearMapper(const LinearMapper&) = default;
+
+      double mInitialValue {};
+      double mSamplesPerPixel {};
+
+      sampleCount operator()(uint32_t column) const noexcept;
+
+      explicit operator bool() const noexcept;
+   };
+   // GCC 9.3.0 fails horribly if you do not initialize variant explicitly here
+   std::variant<LinearMapper, CustomMapper> mMapper { LinearMapper {} };
```

Comment:
The commit that introduces this variant didn't compile for me with clang either.  I fixed it for myself by initializing with one argument `std::in_place_type<LinearMapper>`.  I expect the built fix-ups will be squashed with interactive rebasing in a future version of this PR.

Reply:
I see a commit "Fixes GCC build" adding this line but followed by another that fixes Clang again.

Using `in_place_type` instead is enough to fix Clang in the first commit, without including SampleCount.h.

If you are curious about why the incomlete type was a problem... I looked a bit at the error-novel for clang and I think there is a metaprogram deducing the correct alternative of the variant to initialize from the constructor arguments, which leads to the expansion of a conditional `noexcept` specifier, and somehow the operator () of this type is used in an unevaluated context.

But use `in_place_type` instead and that metaprogram isn't needed to resolve the overloaded constructor.

So, avoiding one more #include inside a header is nice if you can do it, plus there is in-place construction instead of a copy of a LinearMapper.  Small gains.